contracts/CTKNToken.sol

```diff
@@ -0,0 +1,357 @@
+        verified[addr] = hash;
+        VerifiedAddressAdded(addr, hash, msg.sender);
+    }
+
+    /**
+     *  Remove a verified address, and the associated verification hash. If the address is
+     *  unknown to the contract then this does nothing. If the address is successfully removed this
+     *  function must emit `VerifiedAddressRemoved(addr, msg.sender)`.
+     *  It MUST throw if an attempt is made to remove a verifiedAddress that owns Tokens.
+     *  @param addr The verified address to be removed.
+     */
+    function removeVerified(address addr)
+        public
+        onlyOwner
+    {
+        require(balances[addr] == 0);
+        if (verified[addr] != ZERO_BYTES) {
+            verified[addr] = ZERO_BYTES;
+            VerifiedAddressRemoved(addr, msg.sender);
+        }
+    }
+
+    /**
+     *  Update the hash for a verified address known to the contract.
+     *  Upon successful update of a verified address the contract must emit
+     *  `VerifiedAddressUpdated(addr, oldHash, hash, msg.sender)`.
+     *  If the hash is the same as the value already stored then
+     *  no `VerifiedAddressUpdated` event is to be emitted.
+     *  It MUST throw if the hash is zero, or if the address is unverified.
+     *  @param addr The verified address of the person represented by the supplied hash.
+     *  @param hash A new cryptographic hash of the address holder's updated verified information.
+     */
+    function updateVerified(address addr, bytes32 hash)
+        public
+        onlyOwner
+        isVerifiedAddress(addr)
+    {
+        require(hash != ZERO_BYTES);
+        bytes32 oldHash = verified[addr];
+        if (oldHash != hash) {
+            verified[addr] = hash;
+            VerifiedAddressUpdated(addr, oldHash, hash, msg.sender);
+        }
+    }
+
+    /**
+     *  Cancel the original address and reissue the Tokens to the replacement address.
+     *  Access to this function MUST be strictly controlled.
+     *  The `original` address MUST be removed from the set of verified addresses.
+     *  Throw if the `original` address supplied is not a stockholder.
+     *  Throw if the replacement address is not a verified address.
+     *  This function MUST emit the `VerifiedAddressSuperseded` event.
+     *  @param original The address to be superseded. This address MUST NOT be reused.
+     *  @param replacement The address  that supersedes the original. This address MUST be verified.
+     */
+    function cancelAndReissue(address original, address replacement)
+        public
+        onlyOwner
+        isStockholder(original)
+        isNotStockholder(replacement)
+        isVerifiedAddress(replacement)
+    {
+        // replace the original address in the stockholders array
+        // and update all the associated mappings
+        verified[original] = ZERO_BYTES;
+        cancellations[original] = replacement;
+        uint256 holderIndex = holderIndices[original] - 1;
```

Comment:
maybe it should be called holderShares?

Reply:
Not sure if I understand what holderIndices is.  Is it the number of shares for a given a stock holder.

Reply:
Nevermind, I get it.